### PR TITLE
[master] Migrate base/res to new API

### PR DESCRIPTION
**Pull request by rco-odoo**

### CODE REVIEW

File: addons/account/models/account.py [hunk 1]
```diff
@@ -114,7 +114,7 @@ def _check_reconcile(self):
         'account_id', 'tax_id', string='Default Taxes')
     note = fields.Text('Internal Notes')
     company_id = fields.Many2one('res.company', string='Company', required=True,
-        default=lambda self: self.env['res.company']._company_default_get('account.account'))
+        default=lambda self: self.env['res.company']._company_default_get())
     tag_ids = fields.Many2many('account.account.tag', 'account_account_account_tag', string='Tags', help="Optional tags you may want to assign for custom reporting")
```
Comment (rco-odoo) [inline addons/account/models/account.py]: Why did you remove the argument passed to `_company_default_get()` ?

File: openerp/addons/base/res/ir_property.py [hunk 2]
```diff
@@ -2,16 +2,8 @@
...
 
-from openerp import models, api
-from openerp.osv import osv, orm, fields
-from openerp.tools.misc import attrgetter
-from openerp.exceptions import UserError
-
-# -------------------------------------------------------------------------
-# Properties
-# -------------------------------------------------------------------------
+from odoo import api, fields, models, _
```
Comment (rco-odoo) [inline openerp/addons/base/res/ir_property.py]: there is a missing import for `UserError`

File: openerp/addons/base/res/ir_property.py [hunk 3]
```diff
@@ -71,8 +56,8 @@ def _update_values(self, cr, uid, ids, values):
         prop = None
         type_ = values.get('type')
         if not type_:
-            if ids:
-                prop = self.browse(cr, uid, ids[0])
+            if self.exists():
+                prop = self[0]
```
Comment (rco-odoo) [inline openerp/addons/base/res/ir_property.py]: Beware: `self.exists()` returns the subset of `self` that exists; it does not guarantee that all records in `self` exist!
The test `if ids` (`ids` is a nonempty list) should be migrated to `if self` (`self` is a nonempty recordset) in the new API.

File: openerp/addons/base/res/ir_property.py [hunk 4]
```diff
@@ -91,68 +76,68 @@ def _update_values(self, cr, uid, ids, values):
...
-            return record.value_float
-        elif record.type == 'boolean':
-            return bool(record.value_integer)
-        elif record.type == 'integer':
-            return record.value_integer
-        elif record.type == 'binary':
-            return record.value_binary
-        elif record.type == 'many2one':
-            if not record.value_reference:
+        return self.browse(cr, uid, record.id, context).get_by_record()
```
Comment (rco-odoo) [inline openerp/addons/base/res/ir_property.py]: You should make the call through the class, see #10792. Moreover, you should pass `record` itself to get the exact same behavior. It's even simpler!

    @api.v7
    def get_by_record(self, cr, uid, record, context=None):
        return Property.get_by_record(record)

File: openerp/addons/base/res/ir_property.py [hunk 5]
```diff
@@ -91,68 +76,68 @@ def _update_values(self, cr, uid, ids, values):
...
         return False
 
-    def get(self, cr, uid, name, model, res_id=False, context=None):
-        domain = self._get_domain(cr, uid, name, model, context=context)
+    @api.model
+    def get(self, name, model, res_id=False):
+        domain = self._get_domain(name, model)
         if domain is not None:
-            domain = [('res_id', '=', res_id)] + domain
+            domain += [('res_id', '=', res_id)]
```
Comment (rco-odoo) [inline openerp/addons/base/res/ir_property.py]: why not using `append`?

File: openerp/addons/base/res/ir_property.py [hunk 6]
```diff
@@ -91,68 +76,68 @@ def _update_values(self, cr, uid, ids, values):
...
-        if not res:
-            return None
-
-        cid = context.get('force_company')
-        if not cid:
-            company = self.pool.get('res.company')
-            cid = company._company_default_get(cr, uid, model, res[0], context=context)
+    def _get_domain(self, prop_name, model):
+        field = self.env['ir.model.fields'].search([('name', '=', prop_name), ('model', '=', model)], limit=1)
+        if not field: return None
```
Comment (rco-odoo) [inline openerp/addons/base/res/ir_property.py]: Please don't replace the SQL query by an ORM operation. There are good reasons for that query: bypassing access rights and performance.

File: openerp/addons/base/res/ir_property.py [hunk 7]
```diff
@@ -91,68 +76,68 @@ def _update_values(self, cr, uid, ids, values):
...
 
-        return [('fields_id', '=', res[0]), ('company_id', 'in', [cid, False])]
+        company_id = self.env.context.get('force_company', self.env['res.company']._company_default_get().id)
```
Comment (rco-odoo) [inline openerp/addons/base/res/ir_property.py]: Use an `or` in case someone puts `force_company` to `False` in context:

    company_id = self._context.get('force_company') or self.env['res.company']._company_default_get(model).id

File: openerp/addons/base/res/ir_property.py (hunk 4 above)
Comment (rgo-odoo) [inline openerp/addons/base/res/ir_property.py]:
```
property_obj = self.pool['ir.property']
record = property_obj.browse(cr, uid, ids, context=context)
property_obj.get_by_record(cr, SUPERUSER_ID, record, context=context)
```
If we pass `record` itself to new API call then I am afraid we will loose ```cr, uid, context``` values that were used when calling this function from old API.
So, for instance, if executed above code(with say, other than ```admin``` user), the ```api.v7``` version will call the ```api.v8``` with ```record```'s ```uid``` instead of calling it with ```SUPERUSER_ID```

File: openerp/addons/base/res/res_currency.py [hunk 8]
```diff
@@ -1,134 +1,123 @@
...
+    def _compute_current_rate(self):
+        date = self.env.context.get('date') or fields.Datetime.now()
+        company_id = self.env.context.get('company_id', self.env.user._get_company().id)
+        self.env.cr.execute("""SELECT currency_id, rate FROM res_currency_rate
+                           WHERE currency_id in %s
+                             AND name <= %s
+                             AND (company_id is null
+                                 OR company_id = %s)
+                        ORDER BY company_id, name""",
+                                (tuple(self.ids,), date, company_id))
```
Comment (rco-odoo) [inline openerp/addons/base/res/res_currency.py]: Ouch! This query is inefficient: it retrieves **all** rates for the given currencies up to the given date, while only the last ones are useful for the computation... The former query only retrieved one row per currency.

File: openerp/addons/base/res/res_currency.py [hunk 9]
```diff
@@ -1,134 +1,123 @@
...
 
-    @api.cr_uid_records
+    @api.v7
     def round(self, cr, uid, currency, amount):
-        """Return ``amount`` rounded  according to ``currency``'s
+        if isinstance(currency, (int, long)):
+            currency_id = currency
+        else:
+            currency_id = currency.id
+        return self.browse(cr, uid, currency_id).round(amount)
```
Comment (rco-odoo) [inline openerp/addons/base/res/res_currency.py]: don't add more conversion with respect to existing code

File: openerp/addons/base/res/res_currency.py [hunk 10]
```diff
@@ -1,134 +1,123 @@
...
 
-    @api.cr_uid_records
+    @api.v7
     def compare_amounts(self, cr, uid, currency, amount1, amount2):
+        assert currency, "Needs a source currency_id to be compared with."
+        if isinstance(currency, (int, long)):
+            currency_id = currency
+        else:
+            currency_id = currency.id
+        return self.browse(cr, uid, currency_id).compare_amounts(amount1, amount2)
```
Comment (rco-odoo) [inline openerp/addons/base/res/res_currency.py]: don't add more conversion to existing code

File: openerp/addons/base/res/res_currency.py [hunk 11]
```diff
@@ -149,45 +137,51 @@ def compare_amounts(self, cr, uid, currency, amount1, amount2):
...
+    @api.v7
     def is_zero(self, cr, uid, currency, amount):
-        """Returns true if ``amount`` is small enough to be treated as
-           zero according to ``currency``'s rounding rules.
+        assert currency, "Needs a source currency_id to be compared with."
+        if isinstance(currency, (int, long)):
+            currency_id = currency
+        else:
+            currency_id = currency.id
+        return self.browse(cr, uid, currency_id).is_zero(amount)
```
Comment (rco-odoo) [inline openerp/addons/base/res/res_currency.py]: don't add more conversions

File: openerp/addons/base/res/res_currency.py (hunk 8 above)
Comment (rgo-odoo) [inline openerp/addons/base/res/res_currency.py]: I moved query outside loop as I thought it was better to query postgres only once with all ```ids``` than querying for each ```id```

File: openerp/addons/base/res/res_company.py [hunk 12]
```diff
@@ -311,24 +91,242 @@ def _get_header(self,cr,uid,ids):
...
+    vat = fields.Char(related='partner_id.vat', string="Tax ID")
+    company_registry = fields.Char()
+    rml_paper_format = fields.Selection([('a4', 'A4'), ('us_letter', 'US Letter')], string="Paper Format", required=True, default='a4', oldname='paper_format')
+
+    _sql_constraints = [
+        ('name_uniq', 'unique (name)', 'The company name must be unique !')
     ]
+
+    @api.multi
+    def _compute_address(self):
```
Comment (rco-odoo) [inline openerp/addons/base/res/res_company.py]: missing `@api.depends`

File: openerp/addons/base/res/res_company.py [hunk 13]
```diff
@@ -311,24 +91,242 @@ def _get_header(self,cr,uid,ids):
...
+        for company in self.filtered(lambda company: company.partner_id):
+            address_data = company.partner_id.sudo().address_get(adr_pref=['contact'])
+            if address_data['contact']:
+                company.street = company.partner_id.street
```
Comment (rco-odoo) [inline openerp/addons/base/res/res_company.py]: Unless I missed something, here we must use the partner corresponding to `address_data['contact']` instead of `company.partner_id`. BTW, this `address_get` method sucks, I would rather use computed fields like `contact_partner_id`, `delivery_partner_id`, etc.

File: openerp/addons/base/res/res_company.py (hunk 12 above)
Comment (rgo-odoo) [inline openerp/addons/base/res/res_company.py]: Actually I have a question. Will it make any deference if we omit ```depends``` decorator for a non-stored compute method ? Because, we will need this method to be executed each time the compute field is accessed

Comment (rco-odoo) [inline openerp/addons/base/res/res_company.py]: @rgo-odoo in response to your question: omitting the `@api.depends` decorator is not good practice. Non-stored computed fields are kept in cache after they have been computed. Without the decorator, the cache will never be invalidated in the current request, even if what the field depends on has been modified.
